File: Taller_IA.py
import random  # Importamos el módulo random para generar números aleatorios

# Función para simular el lanzamiento de dados
def roll_dice(num_dice=2):
    return [random.randint(1, 6) for _ in range(num_dice)]

# Función para verificar si un número es primo
def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

# Función para manejar el turno de un jugador
def play_turn():
    total = 0
    while True:
        dice = roll_dice()  # Lanzamos 2 dados
        roll_sum = sum(dice)
        total += roll_sum
        print(f"Lanzamiento: {dice}, Suma: {roll_sum}, Total: {total}")

        if total > 23:
            print("Te pasaste de 23. Tu turno termina.")
            return 0  # El jugador pierde si se pasa de 23

        if dice[0] == dice[1]:  # Si saca dobles
            choice = input("Sacaste dobles. ¿Quieres dividir? (s/n): ")
            if choice.lower() == 's':
                print("Dividiendo la mano...")
                hand1 = play_split_hand(dice[0])
                hand2 = play_split_hand(dice[0])
                return max(hand1, hand2)  # Devuelve la mejor puntuación de las dos manos

        if total == 12:
            choice = input("Tienes 12. ¿Quieres lanzar solo un dado? (s/n): ")
            if choice.lower() == 's':
                extra = roll_dice(1)[0]
                total += extra
                print(f"Lanzaste un {extra}. Nuevo total: {total}")
                return total  # Termina el turno después de lanzar un dado adicional

        choice = input("¿Quieres lanzar de nuevo? (s/n): ")
        if choice.lower() != 's':
            return total  # Termina el turno si el jugador no quiere lanzar de nuevo

# Función para manejar una mano dividida
def play_split_hand(initial_value):
    total = initial_value
    print(f"Nueva mano, comenzando con: {total}")
    while True:
        dice = roll_dice(3)  # Lanzamos 3 dados para la mano dividida
        roll_sum = sum(dice)
        total += roll_sum
        print(f"Lanzamiento: {dice}, Suma: {roll_sum}, Total: {total}")

        if total > 23:
            print("Te pasaste de 23 en esta mano dividida.")
            return 0

        choice = input("¿Quieres lanzar de nuevo en esta mano dividida? (s/n): ")
        if choice.lower() != 's':
            return total

# Función para comparar las puntuaciones y determinar el ganador
def compare_scores(scores):
    prime_scores = [score for score in scores if is_prime(score)]
    if prime_scores:
        return max(prime_scores)  # Si hay números primos, el mayor primo gana
    return max(scores)  # Si no hay primos, gana el mayor número

# Función para jugar una ronda completa con todos los jugadores
def play_game(num_players):
    scores = []
    for i in range(num_players):
        print(f"\nTurno del Jugador {i+1}")
        score = play_turn()
        scores.append(score)
        print(f"Puntuación final del Jugador {i+1}: {score}")

    winning_score = compare_scores(scores)
    winners = [i+1 for i, score in enumerate(scores) if score == winning_score]

    print("\nResultados finales:")
    for i, score in enumerate(scores):
        print(f"Jugador {i+1}: {score}")

    if len(winners) == 1:
        print(f"\n¡El Jugador {winners[0]} gana con una puntuación de {winning_score}!")
    else:
        print(f"\n¡Empate entre los Jugadores {', '.join(map(str, winners))} con una puntuación de {winning_score}!")

# Función principal que inicia el juego
def main():
    num_players = int(input("Ingrese el número de jugadores: "))
    while True:
        play_game(num_players)
        play_again = input("¿Quieren jugar otra ronda? (s/n): ")
        if play_again.lower() != 's':
            break

# Aseguramos que el juego solo se inicie si este script es el programa principal
if __name__ == "__main__":
    main()

File: Taller_IA.cpp
#include <iostream>    // Para entrada/salida estándar (cin, cout)
#include <vector>      // Para usar vectores, que son arrays dinámicos
#include <cstdlib>     // Para la función rand() (generación de números aleatorios)
#include <ctime>       // Para la función time() (usada para sembrar el generador de números aleatorios)
#include <algorithm>   // Para funciones como max_element (encontrar el máximo en un vector)

using namespace std;   // Para evitar escribir std:: antes de cin, cout, vector, etc.

// Función para simular el lanzamiento de dados
vector<int> rollDice(int numDice = 2) {
    vector<int> result;
    for (int i = 0; i < numDice; i++) {
        result.push_back(rand() % 6 + 1);
    }
    return result;
}

// Función para verificar si un número es primo
bool isPrime(int n) {
    if (n < 2) return false;
    for (int i = 2; i * i <= n; i++) {
        if (n % i == 0) return false;
    }
    return true;
}

// Función para jugar una mano dividida
int playSplitHand(int initialValue) {
    int total = initialValue;
    cout << "Nueva mano, comenzando con: " << total << endl;
    
    while (true) {
        vector<int> dice = rollDice(3);  // Lanzamos 3 dados para la mano dividida
        int rollSum = 0;
        for (int die : dice) rollSum += die;
        
        total += rollSum;
        cout << "Lanzamiento: ";
        for (int die : dice) cout << die << " ";
        cout << ", Suma: " << rollSum << ", Total: " << total << endl;

        if (total > 23) {
            cout << "Te pasaste de 23 en esta mano dividida." << endl;
            return 0;
        }

        char choice;
        cout << "¿Quieres lanzar de nuevo en esta mano dividida? (s/n): ";
        cin >> choice;
        if (choice != 's' && choice != 'S') return total;
    }
}

// Función para manejar el turno de un jugador
int playTurn() {
    int total = 0;
    while (true) {
        vector<int> dice = rollDice();
        int rollSum = dice[0] + dice[1];
        total += rollSum;
        cout << "Lanzamiento: " << dice[0] << " " << dice[1] << ", Suma: " << rollSum << ", Total: " << total << endl;

        if (total > 23) {
            cout << "Te pasaste de 23. Tu turno termina." << endl;
            return 0;
        }

        if (dice[0] == dice[1]) {  // Si saca dobles
            char choice;
            cout << "Sacaste dobles. ¿Quieres dividir? (s/n): ";
            cin >> choice;
            if (choice == 's' || choice == 'S') {
                cout << "Dividiendo la mano..." << endl;
                int hand1 = playSplitHand(dice[0]);
                int hand2 = playSplitHand(dice[0]);
                return max(hand1, hand2);
            }
        }

        if (total == 12) {
            char choice;
            cout << "Tienes 12. ¿Quieres lanzar solo un dado? (s/n): ";
            cin >> choice;
            if (choice == 's' || choice == 'S') {
                int extra = rollDice(1)[0];
                total += extra;
                cout << "Lanzaste un " << extra << ". Nuevo total: " << total << endl;
                return total;
            }
        }

        char choice;
        cout << "¿Quieres lanzar de nuevo? (s/n): ";
        cin >> choice;
        if (choice != 's' && choice != 'S') return total;
    }
}

// Función para comparar las puntuaciones y determinar el ganador
int compareScores(const vector<int>& scores) {
    vector<int> primeScores;
    for (int score : scores) {
        if (isPrime(score)) primeScores.push_back(score);
    }
    
    if (!primeScores.empty()) {
        return *max_element(primeScores.begin(), primeScores.end());
    }
    return *max_element(scores.begin(), scores.end());
}

// Función para jugar una ronda completa con todos los jugadores
void playGame(int numPlayers) {
    vector<int> scores(numPlayers);
    for (int i = 0; i < numPlayers; i++) {
        cout << "\nTurno del Jugador " << i+1 << endl;
        scores[i] = playTurn();
        cout << "Puntuación final del Jugador " << i+1 << ": " << scores[i] << endl;
    }

    int winningScore = compareScores(scores);
    vector<int> winners;
    for (int i = 0; i < numPlayers; i++) {
        if (scores[i] == winningScore) winners.push_back(i+1);
    }

    cout << "\nResultados finales:" << endl;
    for (int i = 0; i < numPlayers; i++) {
        cout << "Jugador " << i+1 << ": " << scores[i] << endl;
    }

    if (winners.size() == 1) {
        cout << "\n¡El Jugador " << winners[0] << " gana con una puntuación de " << winningScore << "!" << endl;
    } else {
        cout << "\n¡Empate entre los Jugadores ";
        for (size_t i = 0; i < winners.size(); i++) {
            cout << winners[i];
            if (i < winners.size() - 1) cout << ", ";
        }
        cout << " con una puntuación de " << winningScore << "!" << endl;
    }
}

int main() {
    srand(time(0));  // Inicializamos la semilla para números aleatorios
    
    int numPlayers;
    cout << "Ingrese el número de jugadores: ";
    cin >> numPlayers;

    char playAgain;
    do {
        playGame(numPlayers);
        cout << "¿Quieren jugar otra ronda? (s/n): ";
        cin >> playAgain;
    } while (playAgain == 's' || playAgain == 'S');

    return 0;
}
